### PR TITLE
Replace web thing types with capability schemas

### PR DESCRIPTION
**Pull request by benfrancis**
This pull request removes the Web Thing Types section and adds a new Semantic Annotations section to the Web thing Description which explains how to use capability schemas with `@context` and `@type` annotations.

A rendered HTML version is viewable at https://benfrancis.github.io/wot/

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -212,6 +207,107 @@ <h3><code>Event</code> object</h3>
...
+          <p>A <code>@type</code> annotation may also be added to a <a href="#property-object">Property object</a>, <a href="#action-object">Action object</a> or <a href="#event-object">Event object</a> to denote that the property, action or event it defines conforms to a particular schema within the schema repository. These schemas may define data constrains such as data type, unit and expected minimum and maximum values for properties, action inputs and event payloads. The metadata provided in the corresponding <a href="#property-object">Property object</a>, <a href="#action-object">Action object</a> or <a href="#event-object">Event object</a> is expected to conform to those constraints.</p>
+        </section>
+      <div class="example">
+        <div class="example-title marker">
+          Example
+        </div>
+        <pre>{
+  "@context": "https://iot.mozilla.org/schemas/",
+  "@type": ["Light", "OnOffSwitch", "MultiLevelSwitch"],
```
Comment (mrstegeman) [inline index.html]: I encountered this while implementing the capability system:

Since we define both a `LevelProperty` and a `BrightnessProperty`, `Light` and `MultiLevelSwitch` are practically mutually-exclusive. It makes more sense to me for a device to have a `BrightnessProperty` if it's a light than it does to have a `LevelProperty`. If it has a `LevelProperty`, that property is not considered a part of the `Light` capability. However, if it has a `BrightnessProperty`, it cannot be a `MultiLevelSwitch`.

Does that make sense?

File: index.html [hunk 2]
```diff
@@ -212,6 +207,107 @@ <h3><code>Event</code> object</h3>
...
+      "rel": "alternate",
+      "mediaType": "text/html",
+      "href": "/things/lamp"
+    }
+  ]
+}
+</pre>
+      </div>
+    </section>
+    <section>
```
Comment (mrstegeman) [inline index.html]: We should probably make an effort to start using `application/td+json` everywhere.

File: index.html (hunk 1 above)
Comment (benfrancis) [reply index.html]: Good point, I hadn't thought of that.

Of course another possibility is to allow properties to support multiple `@type`s in the same way that things can, so that the same value can be both a BrightnessProperty and a LevelProperty.

```
{
  "name": "My Thing",
  "@context": "https://iot.mozilla.org/schemas",
  "properties": {
    "brightness": {
      "name": "Brightness",
      "@type": ["BrightnessProperty", "LevelProperty"],
      "type": "number",
      "minimum": "0",
      "maximum": "100",
      "unit": "percent"
    }
  }
}
```

This will only work in cases where the constraints of one type are a subset of the other, but it would allow a client which supports LevelProperty but not BrightnessProperty to provide a useful UI.

For now I'll just remove MultiLevelSwitch from the example to keep it simple, and because I realise this isn't something we currently support.

File: index.html (hunk 2 above)
Comment (benfrancis) [reply index.html]: Yes, this is new, but it has made it into the latest W3C Editor's Draft so we should probably start thinking about at least using it for our own Thing Descriptions, although maybe not go as far as only allowing devices to be added which have this media type for their Thing Description.

Note that this is only for the Thing Description itself. There's still an open question as to whether the REST API resources should have their own media type.

**Comment by benfrancis** [comment]
OK, done. I've also updated the example on the schema repository to match.

@hobinjk Any comments before I land this change?

File: index.html [hunk 3]
```diff
@@ -212,6 +207,107 @@ <h3><code>Event</code> object</h3>
...
+          <p>A <code>@type</code> annotation may also be added to a <a href="#property-object">Property object</a>, <a href="#action-object">Action object</a> or <a href="#event-object">Event object</a> to denote that the property, action or event it defines conforms to a particular schema within the schema repository. These schemas may define data constraints such as data type, unit and expected minimum and maximum values for properties, action inputs and event payloads. The metadata provided in the corresponding <a href="#property-object">Property object</a>, <a href="#action-object">Action object</a> or <a href="#event-object">Event object</a> is expected to conform to those constraints.</p>
```
Comment (hobinjk) [inline index.html]: The lack of oxford comma in "data type, unit and expected minimum and maximum values for properties" makes it rather hard to parse.

**Comment by benfrancis** [comment]
OK @hobinjk, I was thinking about your comment about being able to get from this document to a complete annotated Thing Description which the gateway can make best use of.

I've re-organised the Thing Description part of the specification a bit. It now starts with two (rather than one) initial examples. The first is a very simple Thing Description for an On/Off Switch with a single property. The other is a more complex example with actions, events, links and semantic annotations.

I've then moved the `@context` and `@type` members out of their own scary sounding "semantic annotations" section, so they're now just alongside all the other members.

I've also added `@context` and `@type` annotations to the Property, Action and Event object examples to show them in action.

I also took the opportunity to re-write a few parts which have become outdated and include some proper references.

Sorry this is a lot to review at once, but hopefully it's now a bit more obvious how to arrive at a complete Thing Description.

Would you mind re-reviewing? Some of the sentences are maybe still a bit complex, but some of these concepts (e.g. "capability type schemas") are a bit hard to explain concisely.

The updated version is viewable at https://benfrancis.github.io/wot

I agree that the schemas web site could also benefit from more examples, but we can do that separately.
